File: Rust/day1/src/main.rs
use anyhow::{anyhow, Result};

const INPUT: &str = include_str!("../input.txt");

const SPELLED_OUT_NUMBERS: [&str; 10] = [
    "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
];

fn parse_line(line: &str) -> Result<u32> {
    let mut numbers = line.chars().filter_map(|c| c.to_digit(10)).peekable();

    let first = *numbers
        .peek()
        .ok_or_else(|| anyhow!("Line missing a number."))?;
    let last = numbers
        .last()
        .ok_or_else(|| anyhow!("Line missing a number."))?;

    Ok(10 * first + last)
}

fn parse_line_two(line: &str) -> Result<u32> {
    let mut numbers = line
        .char_indices()
        .filter_map(|(i, c)| {
            for (j, number) in SPELLED_OUT_NUMBERS.iter().enumerate() {
                if line[i..].starts_with(number) {
                    return Some(j as u32);
                }
            }

            c.to_digit(10)
        })
        .peekable();

    let first = *numbers
        .peek()
        .ok_or_else(|| anyhow!("Line missing a number."))?;
    let last = numbers
        .last()
        .ok_or_else(|| anyhow!("Line missing a number."))?;

    Ok(10 * first + last)
}

fn solve_part_one(contents: &str) -> Result<u32> {
    contents.lines().map(parse_line).sum()
}

fn solve_part_two(contents: &str) -> Result<u32> {
    contents.lines().map(parse_line_two).sum()
}

fn main() -> Result<()> {
    println!("{}", solve_part_one(INPUT)?);
    println!("{}", solve_part_two(INPUT)?);

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    const TEST_P1: &str = include_str!("../test_input_p1.txt");
    const TEST_P2: &str = include_str!("../test_input_p2.txt");

    #[test]
    fn parsing() {
        assert_eq!(parse_line("1abc2").unwrap(), 12);
        assert_eq!(parse_line("pqr3stu8vwx").unwrap(), 38);
        assert_eq!(parse_line("a1b2c3d4e5f").unwrap(), 15);
        assert_eq!(parse_line("treb7uchet").unwrap(), 77);
    }

    #[test]
    fn part_one() {
        assert_eq!(solve_part_one(TEST_P1).unwrap(), 142);
    }

    #[test]
    fn part_two() {
        assert_eq!(solve_part_two(TEST_P2).unwrap(), 281);
    }
}

File: C/day1/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

const char *numbers[] = {"one", "two",   "three", "four", "five",
                         "six", "seven", "eight", "nine"};

int read_file_into_buffer(char *buffer, FILE *fp) {
  size_t len;
  ssize_t bytes_read = getdelim(&buffer, &len, '\0', fp);

  return bytes_read;
}

int get_number(char *line) {
  int first_number = 0;
  int last_number = 0;

  for (int i = 0; i < strlen(line); i++) {
    if (line[i] >= '0' && line[i] <= '9') {
      if (first_number == 0) {
        first_number = line[i] - '0';
      }
      last_number = line[i] - '0';
    }
  }

  return 10 * first_number + last_number;
}

int starts_with(char *str, const char *prefix) {
  return strncmp(str, prefix, strlen(prefix)) == 0;
}

int get_number_p2(char *line) {
  int first_number = 0;
  int last_number = 0;

  for (int i = 0; i < strlen(line); i++) {
    for (int j = 0; j < 9; j++) {
      if (starts_with(&line[i], numbers[j])) {
        if (first_number == 0) {
          first_number = j + 1;
        }

        last_number = j + 1;
        break;
      }
    }

    if (line[i] >= '0' && line[i] <= '9') {
      if (first_number == 0) {
        first_number = line[i] - '0';
      }
      last_number = line[i] - '0';
    }
  }

  return 10 * first_number + last_number;
}

int solve_part_one(char *input) {
  int total = 0;
  char *line = strtok(input, "\n");

  while (line != NULL) {
    total += get_number(line);

    line = strtok(NULL, "\n");
  }

  return total;
}

int solve_part_two(char *input) {
  int total = 0;
  char *line = strtok(input, "\n");

  while (line != NULL) {
    total += get_number_p2(line);

    line = strtok(NULL, "\n");
  }

  return total;
}

int main() {
  char *buffer = malloc(23000);
  FILE *fp = fopen("test.txt", "r");

  int bytes = read_file_into_buffer(buffer, fp);

  char *copied_string = malloc(23000);

  strcpy(copied_string, buffer);
  printf("Part 1: %d\n", solve_part_one(copied_string));
  free(copied_string);

  printf("Part 2: %d\n", solve_part_two(buffer));
  free(buffer);
}
